### ISSUE TITLE
Optimal settings for CNN model

### ISSUE THREAD

**Issue by patelrajnath** [OpenNMT/OpenNMT-py#693]
I'm trying to train a CNN model to compare with the fairseq-py default model. The training configuration is as follows- 
-gpuid 1 
-encoder_type cnn 
-decoder_type cnn 
-enc_layers 4 
-dec_layers 3 
-rnn_size 256 
-src_word_vec_size 256 
-tgt_word_vec_size 256 
-epochs 30 
-data data-bin_opennmt/en-de 
-save_model cnn-model

I'm using **sdg** as **nag** is not implemented here. The quality scores are quite low (14 BLEU) in comparison to fairseq-py (30 BLEU) default model. **FYI:** I'm using IWSLT14 data set given as an example with fairseq-py repo.  

Please suggest. Thank you.

**Comment by patelrajnath** [comment]
Also, when I tried to use **adam** optimizer, the training ppl doesn't change (same is the case when we train RNN model with adam optimizer)- 
number of parameters: 9216745
encoder:  3317504
decoder:  5899241
Making optimizer for training.
Stage 1: Keys after executing optim.set_parameters(model.parameters())
optim.optimizer.state_dict()['state'] keys: 
optim.optimizer.state_dict()['param_groups'] elements: 
optim.optimizer.state_dict()['param_groups'] element: {'weight_decay': 0, 'eps': 1e-09, 'lr': 0.25, 'amsgrad': False, 'betas': [0.9, 0.999], 'params': [140576663081896, 140576663081824, 140576663082040, 140576663082112, 140576663082184, 140576663082256, 140576663082472, 140576663083480, 140576663083552, 140576663083768, 140576663083912, 140576663083984, 140576663084200, 140576663084344, 140576663084416, 140576663084776, 140576663084632, 140576663084920, 140576663084992, 140576602107976, 140576602108048, 140576602108264, 140576602108408, 140576602108480, 140576602108696, 140576602108840, 140576602108912, 140576602109128, 140576602109272, 140576602109344, 140576602109416, 140576602109488, 140576602109560, 140576602109632, 140576602109704]}

Start training...
 number of epochs: 30, starting from Epoch 1
 batch size: 64

Loading train dataset from data-bin_opennmt/en-de.train.1.pt, number of examples: 146822
/HD_4TB_1/Tools/OpenNMT-py/onmt/modules/ConvMultiStepAttention.py:71: UserWarning: Implicit dimension choice for softmax has been deprecated. Change the call to include dim=X as an argument.
  attn = F.softmax(pre_attn)
/HD_4TB_1/Tools/OpenNMT-py/onmt/Loss.py:144: UserWarning: invalid index of a 0-dim tensor. This will be an error in PyTorch 0.5. Use tensor.item() to convert a 0-dim tensor to a Python number
  return onmt.Statistics(loss[0], non_padding.long().sum(), num_correct)
Epoch  1,    50/ 2295; acc:   0.00; ppl: 26881171418161356094253400435962903554686976.00; xent: 164053.48; 33666 src tok/s; 35276 tgt tok/s;      2 s elapsed
Epoch  1,   100/ 2295; acc:   0.00; ppl: 26881171418161356094253400435962903554686976.00; xent: 1483696.12; 59676 src tok/s; 65140 tgt tok/s;      4 s elapsed
Epoch  1,   150/ 2295; acc:   0.00; ppl: 26881171418161356094253400435962903554686976.00; xent: 2646514.75; 65583 src tok/s; 69603 tgt tok/s;      6 s elapsed

**Comment by srush** [comment]
We haven't really tested the CNN model yet. It was a contribution from a user. I would love for it to be useful, but haven't had the cycles to do the work yet.

**Comment by chenkehai** [comment]
@patelrajnath Could you have a training configuration to get an NMT model to compare with the fairseq-py default model? Could you share the training configuration?

**Comment by patelrajnath** [comment]
Following is the fairseq-py settings-
--lr 0.25 \
--clip-norm 0.1 \
--dropout 0.2 \
--max-tokens 4000 \
--arch fconv_iwslt_de_en

Other parameters are the default.

**Comment by vince62s** [comment]
@hcharaya @patelrajnath 
I fixed a bug introduced by some PR in June. see #1058 
Please check, it shoud be "functional". However I tested it on a small network, giving decent results but far from the transformer. (task DE-EN 4.5 Msegt)

I tested 4 layers of 512 (4 enc 4 dec), adam 0.0002
label_smoothing 0.1
dropout 0.1
position_enconding
batch_size 128

There are probably much better config to test, starting with more layers.

Please let me know how it goes.

**Comment by hcharaya** [comment]
@vince62s I tried training with the latest master including your fix #1058 However i still got very low BLEU score 2.15. I have tried following train and translate commands

python train.py -data iwslt14.tokenized.de-en/germanToEnglish -save_model cnn-model_17_nov -encoder_type cnn -decoder_type cnn -world_size 1 -gpu_ranks 0 -report_every 4 -batch_size 16 -dropout 0.1 -learning_rate 1.0 -max_generator_batches 16 -valid_batch_size 16 -train_steps 600000 -enc_layers 3 -dec_layers 4 -src_word_vec_size 256 -tgt_word_vec_size 256 -rnn_size 256 -optim sgd -log_file log_17_nov.txt

python translate.py -model cnn-model_17_nov_step_250000.pt -src iwslt14.tokenized.de-en/test.de -output pred_250000.txt -replace_unk -verbose -gpu 1 -batch_size 16

perl tools/tokenizer.perl < "pred_250000.txt"> "test_out_bleu_250000.txt"
perl tools/multi-bleu.perl iwslt14.tokenized.de-en/test.en < "test_out_bleu_250000.txt">> "test_out_250000.txt"
[log_17_nov.txt](https://github.com/OpenNMT/OpenNMT-py/files/2593704/log_17_nov.txt)


Attaching the logs for the reference

**Comment by vince62s** [comment]
iwslt is a very small corpus, but I'll try on my end, where did you take the dataset from ?

**Comment by patelrajnath** [comment]
@hcharaya 
As per the log the training is overshooting (PPL is not improving). I would suggest to use **adam 0.0002** instead SGD.

**Comment by hcharaya** [comment]
@vince62s, @patelrajnath  i take the dataset from fairseq german to english (http://opennmt.net/Models-py/) available at 
https://github.com/pytorch/fairseq/blob/e734b0fa58fcf02ded15c236289b3bd61c4cffdf/data/prepare-iwslt14.sh.
 I have tweaked the modification of learning rate and now able to improve BLEU to around 9.7 
So i am trying to do more modifications along these lines to get better results and will inform about latest results tomorrow. 
Thanks for supporting i have only tried SGD till now and in case if i didn't get better results i will try it with adam.

**Comment by hcharaya** [comment]
WIth SGD it didn't converge much so i tried with adam and used following commands for training 
**python3 train.py -data iwslt14.tokenized.de-en/germanToEnglish -save_model cnn-model -encoder_type cnn -decoder_type cnn -world_size 1 -gpu_ranks 0 -report_every 4 -batch_size 16 -dropout 0.1 -learning_rate 0.001 -max_generator_batches 16 -valid_batch_size 16 -train_steps 2000000 -enc_layers 5 -dec_layers 5 -src_word_vec_size 512 -tgt_word_vec_size 512 -rnn_size 512 -optim adam -log_file log_20_nov.txt -reset_optim keep_states -learning_rate_decay 0.99**

by tweaking learning rate at various step i could get 56.25 best validation accuracy and with that got 16.44 BLEU score. Not able to improve further. Any suggestion for improving it further? 
Thanks

**Comment by patelrajnath** [comment]
You can try a few things- 
1. dropout = 0.2
2. positional encoding (quite required for CNN)
3. max grad norm = 1

If it does not help, increase you can add 2-3 more layers or increase kernel_width.

**Comment by hcharaya** [comment]
@patelrajnath I tried 

    1. dropout = 0.2
    2. positional encoding (quite required for CNN)
    3. max grad norm = 1

but i could not achieve validation accuracy of more than 40% while earlier i was able to achieve 56.25 in 40000 steps only. I tried with only positional encoding also but there also i could achieve around 38%.
As you also tried training CNN so what accuracy you were able to achieve?

**Comment by patelrajnath** [comment]
@hcharaya 
It was around 14 (not very good). See the first comment. 

@vince62s What scores did you get with the updated code?

**Comment by vince62s** [comment]
IIRC about 20 or 21 but on WMT dataset, not IWSLT.
